velox/duckdb/conversion/DuckWrapper.cpp

```diff
@@ -242,90 +208,114 @@ double NumericCastToDouble::operation(hugeint_t input) {
   return Hugeint::Cast<double>(input);
 }
 
-template <class T>
-VectorPtr convertDecimalToDouble(
-    ::duckdb::Vector& duckVector,
-    size_t size,
-    const TypePtr& veloxType,
-    memory::MemoryPool* pool,
-    uint8_t scale) {
-  auto* duckData = ::duckdb::FlatVector::GetData<T>(duckVector);
-  auto& duckValidity = ::duckdb::FlatVector::Validity(duckVector);
+inline bool hasValidityVector(::duckdb::Vector& duckVector) {
+  return duckVector.GetType() == LogicalTypeId::HUGEINT ||
+      duckVector.GetType() == LogicalTypeId::TIMESTAMP ||
+      duckVector.GetType() == LogicalTypeId::VARCHAR;
+}
 
-  auto result = BaseVector::create(veloxType, size, pool);
-  auto flatResult = result->as<FlatVector<double>>();
-  double dividend = std::pow(10, scale);
+inline vector_size_t getDuckVectorMaxSize(
+    const ::duckdb::SelectionVector& selection,
+    int32_t size) {
+  // DuckDB vectors doesn't tell what their size is. We are going to use max
+  // index + 1 instead as the vector is guaranteed to be at least that
+  // large.
+  vector_size_t maxIndex = 0;
   for (auto i = 0; i < size; i++) {
-    if (duckValidity.RowIsValid(i)) {
-      double converted =
-          NumericCastToDouble::template operation<T>(duckData[i]) / dividend;
-      flatResult->set(i, converted);
-    } else {
-      result->setNull(i, true);
-    }
+    maxIndex = std::max(maxIndex, (vector_size_t)selection.get_index(i));
   }
-  return result;
+  return (maxIndex + 1);
 }
-
 VectorPtr toVeloxVector(
     int32_t size,
     ::duckdb::Vector& duckVector,
     const TypePtr& veloxType,
-    memory::MemoryPool* pool) {
+    memory::MemoryPool* pool,
+    uint8_t* validityPtr) {
+  VectorPtr veloxFlatVector;
+
+  if (duckVector.GetVectorType() == ::duckdb::VectorType::DICTIONARY_VECTOR) {
```

Comment:
Why are we moving this here from convert?

Reply:
Earlier we had 2 separate implementations convert and convertDecimal so we had moved it out. I can undo this change. But I feel this seems much cleaner where convert only takes care of FlatVectors.

Reply:
I don't like the `validity` vector being part of the `toVeloxVector` API.

> But I feel this seems much cleaner where convert only takes care of FlatVectors.
Why do you feel so?